File: src/test/js/formulas/math/MODTest.ts
import { assert } from "chai";
import { it, describe } from "../../testutils/TestUtils";
import { stub } from "sinon";
import { F7ExceptionName } from "../../../../main/js/errors/F7ExceptionName";
import { NAException } from "../../../../main/js/errors/NAException";
import { ValueException } from "../../../../main/js/errors/ValueException";
import { MOD } from "../../../../main/js/formulas/math/MOD";
import { Grid } from "../../../../main/js/models/common/Grid";
import { CollateralLookupFunction, LookupFunction } from "../../../../main/js/models/common/Types";
import { CommonModels } from "../../testutils/CommonModels";

describe("MOD", function () {
  it("should do normal operations", function () {
    assert.deepEqual(MOD.SELF.run(null, 10, 10), 0);
    assert.deepEqual(MOD.SELF.run(null, 10, 2), 0);
    assert.deepEqual(MOD.SELF.run(null, 10, 3), 1);
    assert.deepEqual(MOD.SELF.run(null, 0, 1628736813.2), 0);
    assert.deepEqual(MOD.SELF.run(null, 218637221.22, 2876.111), 1015.2220000082107);
  });

  it("should do string conversion", function () {
    assert.equal(MOD.SELF.run(null, "10", "3"), 1);
  });

  it("should do pass-through errors", function () {
    assert.deepEqual(MOD.SELF.run(null, 10, new ValueException()), new ValueException());
    assert.deepEqual(MOD.SELF.run(null, new ValueException(), 10), new ValueException());
  });

  it("should use lookup", function () {
    const lookup = stub();
    const collateralLookup = stub();
    const F = new MOD(lookup as LookupFunction, collateralLookup as CollateralLookupFunction);
    collateralLookup.withArgs(CommonModels.A1, CommonModels.M22_RANGE).returns(6);
    collateralLookup.withArgs(CommonModels.A1, CommonModels.G19_RANGE).returns(4);
    assert.deepEqual(F.run(CommonModels.A1, CommonModels.M22_RANGE, CommonModels.G19_RANGE), 2);
    assert.isTrue(collateralLookup.calledWith(CommonModels.A1, CommonModels.M22_RANGE));
    assert.isTrue(collateralLookup.calledWith(CommonModels.A1, CommonModels.G19_RANGE));
    assert.equal(collateralLookup.callCount, 2);
    assert.isTrue(lookup.notCalled);
  });

  it("should handle grids", function () {
    const one = Grid.builder().add(0, 0, 6).add(0, 1, "A").build();
    const two = Grid.builder().add(0, 0, 4).add(0, 1, "B").build();
    assert.deepEqual(MOD.SELF.run(null, one, two), 2);
  });

  it("should return error when argument lengths are wrong", function () {
    assert.deepEqual((MOD.SELF.run(null, "Too few") as NAException).name, F7ExceptionName.NA);
    assert.deepEqual(
      (MOD.SELF.run(null, "A", "B", "Too many") as NAException).name,
      F7ExceptionName.NA
    );
  });
});

File: src/test/java/io/protobase/f7/formulas/math/MODTest.java
package io.protobase.f7.formulas.math;

import io.protobase.f7.errors.DivException;
import io.protobase.f7.errors.NAException;
import io.protobase.f7.errors.ValueException;
import io.protobase.f7.models.Grid;
import io.protobase.f7.testutils.TestFormula;
import org.junit.Test;

import static com.google.common.truth.Truth.assertThat;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.verifyNoMoreInteractions;
import static org.mockito.Mockito.when;

public class MODTest extends TestFormula {
  @Test
  public void testApply_cleanDivision() {
    assertThat(MOD.SELF.apply(null, 10.0, 10.0)).isEqualTo(0.0);
    assertThat(MOD.SELF.apply(null, 10.0, 2.0)).isEqualTo(0.0);
    assertThat(MOD.SELF.apply(null, 10.0, 3.0)).isEqualTo(1.0);
    assertThat(MOD.SELF.apply(null, 0.0, 1628736813.2)).isEqualTo(0.0);
    assertThat(MOD.SELF.apply(null, 218637221.22, 2876.111)).isEqualTo(1015.2220000082107);
  }

  @Test
  public void testApply_stringConversion() {
    assertThat(MOD.SELF.apply(null, "10.0", "10.0")).isEqualTo(0.0);
  }

  @Test
  public void testApply_grid() {
    assertThat(MOD.SELF.apply(null,
        Grid.builder().add(0, 0, 6.0).add(0, 1, "Don't mind me.").build(),
        Grid.builder().add(0, 0, 4.0).add(0, 1, "Don't mind me.").build()
    )).isEqualTo(2.0);
  }

  @Test
  public void test_lookup() {
    MOD F = new MOD(lookup, collateralLookup);
    when(collateralLookup.apply(A1, M22_RANGE)).thenReturn(128.0);
    when(collateralLookup.apply(A1, G19_RANGE)).thenReturn(10.0);
    assertThat(F.apply(A1, M22_RANGE, G19_RANGE)).isEqualTo(8.0);
    verify(collateralLookup).apply(A1, M22_RANGE);
    verify(collateralLookup).apply(A1, G19_RANGE);
    verifyNoMoreInteractions(collateralLookup);
  }

  @Test
  public void testApply_errorsPassThrough() {
    assertThat(MOD.SELF.apply(null, 10.0, new ValueException())).isEqualTo(new ValueException());
    assertThat(MOD.SELF.apply(null, new ValueException(), 10.0)).isEqualTo(new ValueException());
  }

  @Test
  public void testApply_argumentsMismatch() {
    assertThat(MOD.SELF.apply(null, "Too few")).isEqualTo(new NAException());
    assertThat(MOD.SELF.apply(null, "A", "B", "Too many")).isEqualTo(new NAException());
  }

  @Test
  public void testApply_errorDivideByZero() {
    assertThat(MOD.SELF.apply(null, 328467.0, 0.0)).isEqualTo(new DivException());
    assertThat(MOD.SELF.apply(null, 0.0, 0.0)).isEqualTo(new DivException());
  }
}
